fix(angular.copy): should be able to angular.copy Float32Array

DropsOfSerenity opened the pull request:
angular.copy can now copy a Float32Array when only provided src. I added some tests to confirm this.

fixes #10210

DropsOfSerenity committed:
fix(angular.copy): should be able to angular.copy FixedArrays

* angular.copy can now copy a FixedArrays when only provided src.

DropsOfSerenity commented:
@googlebot done.

caitp commented on docs/content/error/ng/cpfad.ngdoc:
```diff
@@ -0,0 +1,7 @@
+@ngdoc error
+@name ng:cpfad
+@fullName Copying Float32Array
```
s/Float32Array/%TypedArray%/g

caitp commented on src/Angular.js:
```diff
@@ -583,6 +582,9 @@ function isPromiseLike(obj) {
 }
 
+function isFloat32Array(value) {
+  return toString.call(value) === '[object Float32Array]';
```
should be `isTypedArray()`, tbh. Using O.p.toString isn't very reliable with ES6 anymore, but we should be able to test it against a regexp like `/^[object (Uint8(Clamped)?)|(Uint16)|(Uint32)|(Int8)|(Int16)|(Int32)|(Float(32)|(64)Array]$`, I think

caitp commented on src/Angular.js:
```diff
@@ -705,12 +707,20 @@ function copy(source, destination, stackSource, stackDest) {
+      } else if (isFloat32Array(source)) {
+        destination = new Float32Array(source.length);
+        for (var j = 0; j < source.length; j++) {
+          destination[j] = source[j];
```
this is not actually necessary, you can just do `destination = new source.constructor(source)` (or `new <specific constructor>(source)` if worried about `constructor` being overwritten)

caitp replied:
(typed array support is great, but this should not be Float32-specific)

petebacondarwin commented:
+1 @caitp 

I was just looking at this too and this is just what I was thinking.
If we are going to support copying Float32Array then we should support all type arrays. Unfortunately IE9 doesn't support them so we need to tread carefully but @caitp's ideas are the direction to go.

@DropsOfSerenity - if you would like to update your PR accordingly that would be great.

- [ ] Update the error doc to be more generic
- [ ] Create a `isTypedArray()` method instead of specific one. Perhaps we could compare `src.constuctor.name` with /(Uint8(Clamped)?)|(Uint16)|(Uint32)|(Int8)|(Int16)|(Int32)|(Float(32)|(64))Array/`
- Update the `copy` method accordingly
- Update the docs

DropsOfSerenity commented:
Excellent, I'll get to it, on a side-note I don't know why the CI build keeps failing, all the tests pass over here.

petebacondarwin commented:
Don't worry. We are having flakiness issues connecting to SauceLabs and BrowserStack when running tests on Travis

DropsOfSerenity commented:
@caitp @petebacondarwin 

Alright thanks for the great advice, I've added a new commit that supports all TypedArrays. I did try using constructor.name instead however that resulted in a memory leak for reasons I don't understand.

One other approach I was thinking of could be to use instanceOf, but then I'd have to add a check to make sure the Type being compared to actually exists (in the case of ie9- it doesn't)

Either way this solution seems to be working well, updated the tests to check all TypedArrays.

DropsOfSerenity commented:
@caitp sweet looks like CI finally passed :)

caitp commented:
party =) I'll take a look at this later today, gotta finish up a quick patch I'm writing first

caitp commented on docs/content/error/ng/cpta.ngdoc:
```diff
@@ -0,0 +1,7 @@
+@ngdoc error
+@name ng:cpfad
```
should be `@name ng:cpta`

caitp commented on src/Angular.js:
```diff
@@ -583,6 +582,9 @@ function isPromiseLike(obj) {
 }
 
+function isTypedArray(value) {
+  return /^\[object (Uint8(Clamped)?)|(Uint16)|(Uint32)|(Int8)|(Int16)|(Int32)|(Float(32)|(64))Array\]$/.test(toString.call(value));
```
It's not going to matter to code-gen (at least in v8), but we should probably declare the regexp as a variable outside of the function, so that it's more readable

caitp commented on test/AngularSpec.js:
```diff
@@ -78,6 +78,42 @@ describe('angular', function() {
+    it("should throw an exception if a TypedArray is the destination", function() {
+      if (typeof Float32Array !== 'undefined') {
+        var arraysToCheck = [
+          Uint8Array, Uint16Array, Uint32Array, Int8Array, Int16Array,
```
we should actually split these into multiple `it()` blocks... @matsko has a tool for doing this in https://github.com/angular/angular.js/pull/10864 but it has not landed yet. Matias do you think you could get that in soon or else submit a followup PR to cleanup these tests if that lands first?

DropsOfSerenity replied:
@caitp I can split these, it's not big deal, don't wanna commit code that would have to be cleaned up later :)

caitp commented:
I lied, I guess I did review the PR before finishing my patch :3 It looks good to me, just fix up the nits and squash the commits

DropsOfSerenity commented:
@caitp Ok i squashed with my changes. :3

caitp commented:
good stuff, looks ready to merge to me

caitp commented:
i guess this was really a feature rather than a fix >_- oh well, not changing that message :>

Hypercubed commented:
This change is great... have you considered exposeding `istypedArray` as `angular.isTypedArray`?

caitp commented:
glad you like it @Hypercubed! but... exposing these utilities is problematic in many ways =\ i don't think it would be smart... Besides, `@@toStringTag` makes it unreliable anyhow
